stl/inc/algorithm

```diff
@@ -4626,6 +4626,155 @@ _FwdIt2 rotate_copy(_ExPo&&, _FwdIt1 _First, _FwdIt1 _Mid, _FwdIt1 _Last, _FwdIt
 }
 
+#ifdef __cpp_lib_concepts
+namespace ranges {
+    // VARIABLE ranges::rotate
+    class _Rotate_fn : private _Not_quite_object {
+    public:
+        using _Not_quite_object::_Not_quite_object;
+
+        template <permutable _It, sentinel_for<_It> _Se>
+        _NODISCARD constexpr subrange<_It> operator()(_It _First, _It _Mid, _Se _Last) const {
+            _Adl_verify_range(_First, _Mid);
+            _Adl_verify_range(_Mid, _Last);
+            auto _UResult = _Rotate_unchecked(
+                _Get_unwrapped(_STD move(_First)), _Get_unwrapped(_STD move(_Mid)), _Get_unwrapped(_STD move(_Last)));
+
+            return _Rewrap_subrange<subrange<_It>>(_First, _STD move(_UResult));
+        }
+
+        // clang-format off
+        template <forward_range _Rng>
+            requires permutable<iterator_t<_Rng>>
+        _NODISCARD constexpr borrowed_subrange_t<_Rng> operator()(_Rng&& _Range, iterator_t<_Rng> _Mid) const {
+            _Adl_verify_range(_Ubegin(_Range), _Mid);
+            auto _UResult = _Rotate_unchecked(_Ubegin(_Range), _Get_unwrapped(_STD move(_Mid)), _Uend(_Range));
+
+            return _Rewrap_subrange<borrowed_subrange_t<_Rng>>(_Range, _STD move(_UResult));
+        }
+        // clang-format on
+    private:
+        template <class _It, class _Se>
+        _NODISCARD static constexpr subrange<_It> _Rotate_unchecked(_It _First, _It _Mid, _Se _Last) {
+            // Exchange the ranges [_First, _Mid) and [_Mid, _Last)
+            // that is, rotates [_First, _Last) left by distance(_First, _Mid) positions
+            _STL_INTERNAL_STATIC_ASSERT(permutable<_It>);
+            _STL_INTERNAL_STATIC_ASSERT(sentinel_for<_Se, _It>);
+
+            auto _End = _Get_final_iterator_unwrapped<_It>(_First, _STD move(_Last));
+            if (_First == _Mid) {
+                return {_End, _End};
+            }
+
+            if (_Mid == _End) {
+                return {_STD move(_First), _STD move(_End)};
+            }
+
+            if constexpr (random_access_iterator<_It>) {
+                _RANGES _Reverse_common(_First, _Mid);
+                _RANGES _Reverse_common(_Mid, _End);
+                _RANGES _Reverse_common(_First, _End);
+                _First += (_End - _Mid);
+                return {_STD move(_First), _STD move(_End)};
+            } else if constexpr (bidirectional_iterator<_It>) {
+                _RANGES _Reverse_common(_First, _Mid);
+                _RANGES _Reverse_common(_Mid, _End);
+                while (_First != _Mid && _End != _Mid) {
+                    _RANGES iter_swap(_First, --_End);
+                    ++_First;
+                }
+                if (_First == _Mid) {
+                    _RANGES _Reverse_common(_Mid, _End);
+                    return {_STD move(_First), _STD move(_End)};
+                } else {
+                    _RANGES _Reverse_common(_First, _Mid);
+                    return {_STD move(_First), _STD move(_End)};
+                }

```

Comment:
This is equivalent to:
```c++
_RANGES _Reverse_common(_First, _End);
return {_STD move(_First), _STD move(_End)};
```
The return value seems wrong to me: (1) `_End` is no longer at the end of the range since we've been decrementing it in the loop above, and (2) `_First` is in the right place only if we exited the while loop because `_End == _Mid`.

Reply:
Yeah genius me fixed this on one machine without pushing the changes and mentally marked it as solved. Now I have to resolve some merge conflicts